### PR TITLE
TASK: Update CustomStylingForCKEditor to CKEditor5

### PR DESCRIPTION
**Pull request by Stefan-Kosker**
Examples fitted to CKEditor5 for the sake of documentation.

### CODE REVIEW

File: Resources/Private/CustomStylingForCkEditor/package.json [hunk 1]
```diff
@@ -7,10 +7,12 @@
...
   "devDependencies": {
-    "@neos-project/neos-ui-extensibility": "~1.0.8",
-    "@neos-project/react-ui-components": "~1.0.8"
+    "@neos-project/neos-ui-extensibility": "^1.3.0"
   },
   "neos": {
     "buildTargetDirectory": "../../Public/CustomStylingForCkEditor"
+  },
+  "dependencies": {
+    "@ckeditor/ckeditor5-core": "^11.0.0"
```
Comment (dimaip) [inline Resources/Private/CustomStylingForCkEditor/package.json]: This shouldn't be needed

File: Resources/Private/CustomStylingForCkEditor/src/manifest.js [hunk 2]
```diff
@@ -1,25 +1,28 @@
...
-        hoverStyle: 'brand'
-    });
-
-    const formattingRules = ckEditorRegistry.get('formattingRules');
-
-   formattingRules.set('Neos.Neos.Ui.ExtensibilityExamples:MyCustomSpan', {
-       style: {element: 'span', attributes: {style: 'background-color: red'}},
-       config: formattingRules.config.addToExtraAllowedContent('span[style]')
-    });
+    richtextToolbar.set('splitAdd', {
```
Comment (dimaip) [inline Resources/Private/CustomStylingForCkEditor/src/manifest.js]: splitAdd?

File: Resources/Private/CustomStylingForCkEditor/src/manifest.js [hunk 3]
```diff
@@ -1,25 +1,28 @@
...
+        commandName: 'neosSplit',
```
Comment (dimaip) [inline Resources/Private/CustomStylingForCkEditor/src/manifest.js]: neosSplit?

File: Resources/Private/CustomStylingForCkEditor/src/examplePlugin.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import Command from '@ckeditor/ckeditor5-core/src/command';
+
+class ExampleCommand extends Command {
+    execute(callback) {
+        const model = this.editor.model;
+        let ranges = model.document.selection._selection._ranges;
```
Comment (drotalion) [inline Resources/Private/CustomStylingForCkEditor/src/examplePlugin.js]: I am not sure, if you would get the ranges like this @dimaip . I had no chance using getRanges() (always returns empty generator for me). You can directly access to selection and ranges with this . Check `selection.js` in `ckeditor5-engine/model`for further info

File: Resources/Private/CustomStylingForCkEditor/src/examplePlugin.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        ranges.forEach(range => {
+            model.change(writer => {
+                writer.setAttribute('exampleAttribute', true, range);
```
Comment (drotalion) [inline Resources/Private/CustomStylingForCkEditor/src/examplePlugin.js]: I know, you can achieve this more easily with an attributeCommand. I can do this, but then, I would suggest, I will create another example showing how a command controller gets called. I think, commandControllers will be used at least as often as attributeCommand.

File: Resources/Private/CustomStylingForCkEditor/src/ExampleButton.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+@connect($transform({
+    fusionPath: $get('cr.nodes.focused.fusionPath')
+}), {
+    persistChanges: actions.Changes.persistChanges,
+    commenceNodeCreation: actions.CR.Nodes.commenceCreation
+})
+@neos(globalRegistry => ({
+    i18nRegistry: globalRegistry.get('i18n')
+}))
+export default class SpellCheckButton extends PureComponent {
```
Comment (drotalion) [inline Resources/Private/CustomStylingForCkEditor/src/ExampleButton.js]: SpellCheckButton?

**Comment by drotalion** [comment]
Btw, the reference appears to be wrong:
https://github.com/neos/neos-development-collection/pull/2157

**Comment by skurfuerst** [comment]
@dimaip anything we can do to get this merged? :-) All the best, Sebastian

File: Resources/Private/CustomStylingForCkEditor/package.json [hunk 7]
```diff
@@ -7,10 +7,12 @@
     "watch": "neos-react-scripts watch"
   },
   "devDependencies": {
-    "@neos-project/neos-ui-extensibility": "~1.0.8",
-    "@neos-project/react-ui-components": "~1.0.8"
+    "@neos-project/neos-ui-extensibility": "^1.3.0"
```
Comment (dimaip) [inline Resources/Private/CustomStylingForCkEditor/package.json]: I guess the version constraint should be update to the latest release, right?

File: Resources/Private/CustomStylingForCkEditor/src/manifest.js [hunk 8]
```diff
@@ -1,25 +1,28 @@
 import manifest from '@neos-project/neos-ui-extensibility';
-import {IconButton} from '@neos-project/react-ui-components';
+import {$add, $get} from 'plow-js';
+import ExamplePlugin from './examplePlugin';
+import ExampleButton from './ExampleButton';
 
-manifest('main', {}, globalRegistry => {
-    const ckEditorRegistry = globalRegistry.get('ckEditor');
+const addPlugin = (Plugin, isEnabled) => (ckEditorConfiguration, options) => {
```
Comment (dimaip) [inline Resources/Private/CustomStylingForCkEditor/src/manifest.js]: Would be nice to expose this helper function via extensibility api, so it wouldn't be necessary to copy-paste it all the time. But can be done later ofc.

File: Resources/Private/CustomStylingForCkEditor/package.json (hunk 7 above)
Comment (dimaip) [reply Resources/Private/CustomStylingForCkEditor/package.json]: Or just leave `*` and leave a comment to use whatever current version.

Comment (dimaip) [reply Resources/Private/CustomStylingForCkEditor/package.json]: Ah, yeah, it's not possible to leave comments here.

**Comment by dimaip** [comment]
@skurfuerst while we are at it, we could perhaps start thinking how to simplify this API for most common tasks (e.g. using custom classes), perhaps even creating a higher-level API for this at NodeTypes.yaml level.

**Comment by skurfuerst** [comment]
@dimaip agreed, though I have to move that thinking to after-neos-conference ;)

**Pull request merged**